stock_picking_group_by_partner_by_carrier/models/stock_move.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2020 Camptocamp (https://www.camptocamp.com)
+# Copyright 2020-2021 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from collections import namedtuple
+from itertools import groupby
+
+from odoo import api, fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    # store the first group the move was in when created, used to keep track of
+    # original group's name when creating a joint group for merged transfers,
+    # and for cancellation of a sales order (to cancel only the moves related
+    # to it)
+    original_group_id = fields.Many2one(
+        comodel_name="procurement.group",
+        string="Original Procurement Group",
+    )
+
+    @api.model
+    def _prepare_merge_moves_distinct_fields(self):
+        # Prevent merging pulled moves. This allows to cancel a SO without
+        # canceling pulled moves from other SO as we ensure they are not
+        # merged.
+        return super()._prepare_merge_moves_distinct_fields() + ["original_group_id"]
+
+    def _assign_picking(self):
+        result = super(
+            StockMove, self.with_context(picking_no_overwrite_partner_origin=1)
+        )._assign_picking()
+        return result
+
+    def _assign_picking_post_process(self, new=False):
+        moves_by_picking = groupby(
+            sorted(self, key=lambda m: m.picking_id.id), key=lambda m: m.picking_id
+        )
+        for picking, imoves in moves_by_picking:
+            merged = picking._merge_procurement_groups()
+            if merged:
+                moves = self.browse(m.id for m in imoves)
+                moves.picking_id._update_merged_origin()
+                moves._on_assign_picking_message_link()
+        res = super()._assign_picking_post_process(new=new)
+        return res
+
+    def _on_assign_picking_message_link(self):
+        sales = self.sale_line_id.order_id
+        if sales:
+            self.picking_id.message_post_with_view(
+                "mail.message_origin_link",
+                values={"self": self.picking_id, "origin": sales, "edit": True},
+                subtype_id=self.env.ref("mail.mt_note").id,
+            )
+
+    def _search_picking_for_assignation_domain(self):
+        domain = super()._search_picking_for_assignation_domain()
+        if (
+            not self.picking_type_id.group_pickings
+            or self.partner_id.disable_picking_grouping
+            or self.group_id.sale_id.picking_policy == "one"
+        ):
+            return domain
+
+        # remove group
+        domain = [x for x in domain if x[0] != "group_id"]
+
+        grouping_domain = self._assign_picking_group_domain()
+
+        return domain + grouping_domain
+
+    # TODO: this part and everything related to generic grouping
```

Comment:
@simahawk @jbaudoux What do you think of this ?

How choosing the good domain as I suppose several group_by_... modules can be installed together

Reply:
Yeah, you can have more than one. So far, with the ones available today we didn't have problems.
If you need to take full control, you should probably override this in your custom module.
The idea here is to have a base module for defining an hook and some sort of api for this feat.